webapp/channels/src/components/profile_popover/profile_popover_call_button_wrapper/index.tsx

```diff
@@ -51,6 +53,10 @@ const CallButton = ({
 
     const isCallsEnabled = useSelector((state: GlobalState) => getIsCallsEnabled(state));
     const dmChannel = useSelector((state: GlobalState) => getChannelByName(state, getDirectChannelName(currentUserId, userId)));
+    const callsConfig = useSelector(getCallsConfig);
+    const currentUser = useSelector((state: GlobalState) => getUser(state, currentUserId));
+    const callsExplicitlyDisabled = useSelector((state: GlobalState) => callsChannelExplicitlyDisabled(state, dmChannel?.id ?? ''));
+    const callsExplicitlyEnabled = useSelector((state: GlobalState) => callsChannelExplicitlyEnabled(state, dmChannel?.id ?? ''));
```

Comment:
You are getting all these values just to decide the if. This could be potentially improved by using a single selector for the single value we need (`shouldRenderButton` or something like that).

So... this could end up something like this:

```suggestion
    const shouldRenderButton = useSelector((state: GlobalState) => {
        if (!isCallsEnabled) {
            return false;
        }
        if (callsChannelExplicitlyDisabled(state, dmChannel?.id ?? '')) {
            return false;
        }
       if (isSystemAdmin(getCurrentUser(state))?.roles) {
           return true;
       }
       if (getCallsConfig(state).DefaultEnabled) {
           return true;
       }
       if (callsChannelExplicitlyEnabled(state, dmChannel?.id ?? '')) {
           return true;
       }
       
       return false;
   }
```

This way only one selector is run on every state change, and hopefully it is able to early return more often. Also, the value it is returning is a boolean and nearer the business logic, so there will be even less re-renders. Finally, doing this `if (x) return A; if (y) return B; return C` pattern makes it a bit easier to read (I hope) and may not need as many comments.

Reply:
Now that I look deeper, you removed all the calls to `isCallsEnabled` apart of what would end up in this selector, so you can move also that logic here, so there is one less state change that may trigger a re-render and one less selector to run when the state changes.

Reply:
And... if we move the logic to get the dmChannel down to ProfilePopoverCallButton (or even to a new function `ensureDirectChannel` that would substitute the call to `createDirectChannel` there), you could also get the `dmChannel` selector inside this new selector, so superfluous changes in the channel object don't affect the rendering of this button.

Reply:
Thanks @larkox, the single selector makes total sense.

> Now that I look deeper, you removed all the calls to isCallsEnabled apart of what would end up in this selector, so you can move also that logic here, so there is one less state change that may trigger a re-render and one less selector to run when the state changes.

We are still using that in `hasDMCall` :thinking: 

> And... if we move the logic to get the dmChannel down to ProfilePopoverCallButton (or even to a new function ensureDirectChannel that would substitute the call to createDirectChannel there), you could also get the dmChannel selector inside this new selector, so superfluous changes in the channel object don't affect the rendering of this button.

Considering changes to the channel object would need to happen while the profile popover is open is this really worth the refactoring?

Reply:
If I remember correct (I may be wrong) we store in the channel object the date of the last message. So any new post in the channel would re-render this. Similar with some websocket events, as the channel name changing, or stuff like that.

The reality is that right now it is not a bottleneck, so it is not urgent to have it "correct". I am fine leaving as is if you deem it too much effort at the moment. It is just one of those things that "since we saw it, we can fix it". If we don't fix it now, it will continue unnoticed for a long time probably :P But again... it is nothing critical right now.

Reply:
> We are still using that in hasDMCall

Oh, didn't see that. Sometimes github review view gets me a bit of a tunnel vision :sweat_smile:

Reply:
> So any new post in the channel would re-render this. Similar with some websocket events, as the channel name changing, or stuff like that.

That's a good point. Given it's a single DM we are considering here, I think we should be relatively safe though as I wouldn't expect much throughput.